plutus-core/plutus-ir/src/PlutusIR/Transform/NonStrict.hs

```diff
@@ -28,25 +28,50 @@ Since we are constructing a global substitution, so we need globally unique
 names to avoid clashes.
 -}
 
+{- Note [Using unit versus force/delay]
+We have force/delay in PIR, so we can use that for non-strict bindings, which is quite nice.
+
+However, we retain the *option* to use unit-lambdas instead, since we rely on this pass to
+handle recursive, non-function bindings and give them function types. `delayed x` is not a
+function type but `() -> x` is!
```

Comment:
> Given this, can we have some optimization here as well?

In some cases. That would let us handle polymorphic functions, but not recursive values which are not functions! An edge case, true, but one that Haskell supports, and that this trick also handles. There's a test for it too.

> By the way, does this transformation change the semantics of a program?

No. PIR has non-strict let bindings, this turns them into strict let-bindings that behave the same.

I think what you're asking is "what do we do with Haskell let-bindings?". The answer is that we turn them into non-strict let-bindings unless we can tell the RHS is a value. So we do actually preserve some of the non-strictness of Haskell, rather than interpreting it strictly. I don't 100% remember why I decided to do this, but I think it was hard to avoid random things unexpectedly breaking if you didn't

Reply:
Thanks, makes perfect sense.